targets/TARGET_Freescale/TARGET_KLXX/TARGET_KL46Z/flash_api.c

```diff
@@ -0,0 +1,225 @@
+    .flash_start = 0x0,
+    .flash_size = 0x40000,
+    .sectors = sectors_info,
+};
+
+/*
+    This binary blob (thumb code) sets r9 (static base) as the code we are jumping to
+    is PIC (position independent code).
+
+    These are the instructions (r0 is a pointer to arg_t):
+        push {r5, lr, r4}
+        mov r5, r9
+        push {r5}
+        ldr r5, [r0, #20]
+        ldr r3, [r0, #16]
+        mov r9, r3
+        ldr r3, [r0, #12]
+        ldr r2, [r0, #8]
+        ldr r1, [r0, #4]
+        ldr r0, [r0, #0]
+        blx r5
+        pop {r5}
+        mov r9, r5
+        pop {r4-r5, pc}
+        bx r14
+*/
+static uint32_t jump_to_flash_algo[] = {
+    0x464DB530,
+    0x6945B420,
+    0x46996903,
+    0x688268C3,
+    0x68006841,
+    0xBC2047A8,
+    0xBD3046A9
+};
+
+static uint32_t get_sector_index(flash_t *obj, uint32_t address)
+{
+    // check where address belongs to
+    size_t sector_index = 0;
+    size_t sectors_count = sizeof(obj->target_config->sectors) / sizeof(sector_info_t);
+    for (; sector_index < sectors_count; sector_index++) {
+        if ((address > obj->target_config->sectors[sector_index].start) && 
+            (address < (obj->target_config->sectors[sector_index].start +obj->target_config->sectors[sector_index].size))) {
+            break;
+        }
+    }
+    return sector_index;
+}
+
+int32_t flash_init(flash_t *obj)
+{
+    core_util_critical_section_enter();
+
+    obj->flash_algo = &flash_algo_config;
+    obj->target_config = &flash_target_config;
+
+    args_t arguments = {
+        .r0 = 0,
+        .r1 = 0,
+        .r2 = 0,
+        .r3 = 0,
+        .r9 = (uint32_t)obj->flash_algo->algo_blob + obj->flash_algo->static_base,
+        .pc = (uint32_t)obj->flash_algo->algo_blob + obj->flash_algo->init
+    };
+    int32_t ret = ((flash_algo_jump_t)(((uint32_t)&jump_to_flash_algo) | 1))(&arguments);
+
+    core_util_critical_section_exit();
+    return ret;
+}
+
+int32_t flash_free(flash_t *obj)
+{
+    // uninit is not called here as it deinitializes MCU
+    return 0;
+}
+
+
+int32_t flash_erase_sector(flash_t *obj, uint32_t address)
+{
+    size_t sector_index = get_sector_index(obj, address);
+
+    // erase sector boundary
+    if ((address % obj->target_config->sectors[sector_index].size) != 0) {
+        return -1;
+    }
+
+    core_util_critical_section_enter();
+
+    args_t arguments = {
+        .r0 = address,
+        .r1 = 0,
+        .r2 = 0,
+        .r3 = 0,
+        .r9 = (uint32_t)obj->flash_algo->algo_blob + obj->flash_algo->static_base,
+        .pc = (uint32_t)obj->flash_algo->algo_blob + obj->flash_algo->erase_sector
+    };
+    int32_t ret = ((flash_algo_jump_t)(((uint32_t)&jump_to_flash_algo) | 1))(&arguments);
+
+    core_util_critical_section_exit();
+    return ret;
+}
+
+
+int32_t flash_program_page(flash_t *obj, uint32_t address, const uint8_t *data, uint32_t size)
+{ 
+    // write size boundary
+    if ((address % obj->target_config->page_size) != 0) {
```

Comment:
You could probably just assert here, as the application is doing something wrong.